CMSgov/mint-app
# Multiuser Task List Locking Subscription

# EASI-1995

## Changes and Description

- RefCounted implementation of simple task list locking.
- Note: Nothing in this PR actually secures the transactions, that will be done as a separate pass for simplicity.



## How to test this change

### Create a New Websocket Request
![image](https://user-images.githubusercontent.com/100007843/179571524-f3f2767d-8814-410d-b83f-0abd0a98e85a.png)

### Create a GQL_CONNECTION_INIT Request
![image](https://user-images.githubusercontent.com/100007843/179571624-b134e6da-25c5-49cd-a677-4bd6aebe3d27.png)

```
{
    "type": "connection_init"
}
```

### Create an onTaskListSectionsChanged Request
![image](https://user-images.githubusercontent.com/100007843/179571792-05ad5276-5ece-42de-bce5-c2b5471e514a.png)

```
{
    "payload": {
        "operationName": "onTaskListSectionLocksChanged",
        "query": "subscription onTaskListSectionLocksChanged($modelPlanID: UUID!) { onTaskListSectionLocksChanged(modelPlanID: $modelPlanID) { changeType lockStatus { modelPlanID section lockedBy refCount } } }",
        "variables": {
            "modelPlanID": "f11eb129-2c80-4080-9440-439cbe1a286f"
        }
    },
    "type": "start"
}
```

### POST: LockTaskListSection

#### Query
```
mutation lockTaskListSection($modelPlanID: UUID!, $section: TaskListSection!)
{
    lockTaskListSection(modelPlanID: $modelPlanID, section: $section)
}
```

#### GraphQL Variables
```
{
  "modelPlanID": "f11eb129-2c80-4080-9440-439cbe1a286f",
  "section": "MODEL_BASICS"
}
```

### POST: UnlockTaskListSection

#### Query
```
mutation unlockTaskListSection($modelPlanID: UUID!, $section: TaskListSection!)
{
    unlockTaskListSection(modelPlanID: $modelPlanID, section: $section)
}
```

#### GraphQL Variables
```
{
  "modelPlanID": "f11eb129-2c80-4080-9440-439cbe1a286f",
  "section": "MODEL_BASICS"
}
```

### Testing Steps
- Run the GQL_CONNECTION_INIT request to prime your connection
  - Your response should look like this ('ka' type messages are KeepAlive messages, you will receive these at regular intervals) 
![image](https://user-images.githubusercontent.com/100007843/179572068-fe6a9f9f-7354-46e4-81b4-b082ae0835fb.png)
- Run the onTaskListSectionLocksChanged request to subscribe to task list section lock changes
  - Your response should look like this
![image](https://user-images.githubusercontent.com/100007843/179572282-5a741ccd-c204-4a8c-a4f4-0f65d2fc3fca.png)
- Lock a task list section
  - Your response in the WebSocket tab should look like this
![image](https://user-images.githubusercontent.com/100007843/179572450-d466fe04-45f3-4b3f-b123-ab51eeb5459c.png)
- Unlock a task list section
  - Your response in the WebSocket tab should look like this
![image](https://user-images.githubusercontent.com/100007843/179572618-a79d2701-59fd-4941-a950-1ae38cb5d78b.png)
- Repeat locking and unlocking until you are satisfied
  - Bear in mind that with refcounting, if you lock a task list section twice you MUST unlock it twice to release the lock. This is by design, and is guarded by WebSocket disconnects where not explicit in order to support multiple tabs.


## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
